airbyte-integrations/connectors/source-trello/source_trello/source.py

```diff
@@ -0,0 +1,227 @@
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+import time
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+
+import requests
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import HttpAuthenticator
+
+
+class TrelloStream(HttpStream, ABC):
+    url_base = "https://api.trello.com/1/"
+
+    # Define primary key as sort key for full_refresh, or very first sync for incremental_refresh
+    primary_key = "id"
+
+    # Page size
+    limit = None
+
+    extra_params = None
+
+    # https://developer.atlassian.com/cloud/trello/guides/rest-api/rate-limits/
+    queries_per_900_seconds = 100  # 100 queries per 900 seconds
+
+    def __init__(self, config: Mapping[str, Any]):
+        super().__init__(authenticator=config["authenticator"])
+        self.start_date = config["start_date"]
+        self._config = config
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = {"limit": self.limit, "since": self.start_date}
+        if next_page_token:
+            params.update(**next_page_token)
+        if self.extra_params:
+            params.update(self.extra_params)
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        json_response = response.json()
+        for record in json_response:
+            yield record
+
+        time.sleep(900 / self.queries_per_900_seconds)
+
+
+class ChildStreamMixin:
+    parent_stream_class: Optional[TrelloStream] = None
+
+    def stream_slices(self, sync_mode, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        for item in self.parent_stream_class(config=self._config).read_records(sync_mode=sync_mode):
+            yield {"id": item["id"]}
+
+        yield from []
+
```

Comment:
is this mixin class? if true, its variables can't be used/modified by other classes and a mixin instance can't "protected" variables (self._config etc) of others. Maybe will we add intermediate parent class for this functionality?

Reply:
Yes, for mixin it not forbidden. Mixin is a language concept that allows a programmer to inject some code into a class. And that's all. 

Why ChildStreamMixin is the only mixin? Because ChildStreamMixin injects stream slice logic using parent class.

Reply:
Yep, it is recommendation only: Mixin shouldn't use internal logic of child classes(using of protected methods/variables) and a child class shouldn't know about its settings. And the main reason of its using is case where you want to add a new same logic for different classes without a common parent. Classic inheritance if more than enough for this case.

Reply:
Updated